src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp

```diff
@@ -0,0 +1,1104 @@
+		}
+		_lidarFault = 1;
+	// disable fault if ok
+	} else if (_lidarFault) {
+		_lidarFault = 0;
+		mavlink_log_info(_mavlink_fd, "[lpe] lidar OK");
+		warnx("[lpe] lidar OK");
+	}
+
+	// kalman filter correction if no fault
+	// want to ignore corrections > 3 std. dev since lidar gives
+	// bogus readings at times
+	if (_lidarFault == 0) {
+		math::Matrix<n_x, n_y_lidar> K = _P*C.transposed()*S_I;
+		_x += K*r;
+		_P -= K*C*_P;
+	}
+	_time_last_lidar = _sub_distance.get().timestamp;
+}
+
+void BlockLocalPositionEstimator::correctGps() {	// TODO : use another other metric for glitch detection
+
+	// gps measurement in local frame
+	double  lat = _sub_gps.get().lat*1.0e-7;
+	double  lon = _sub_gps.get().lon*1.0e-7;
+	float  alt = _sub_gps.get().alt*1.0e-3f;
+
+	float px = 0;
+	float py = 0;
+	float pz = alt - _gpsAltHome;
+	map_projection_project(&_map_ref, lat, lon, &px, &py);
+
+	//printf("gps: lat %10g, lon, %10g alt %10g\n", lat, lon, double(alt));
+	//printf("home: lat %10g, lon, %10g alt %10g\n", _sub_home.lat, _sub_home.lon, double(_sub_home.alt));
+	//printf("local: x %10g y %10g z %10g\n", double(px), double(py), double(pz));
+
+	math::Vector<6> y;
+	y(0) = px;
+	y(1) = py;
+	y(2) = pz;
+	y(3) = _sub_gps.get().vel_n_m_s;
+	y(4) = _sub_gps.get().vel_e_m_s;
+	y(5) = _sub_gps.get().vel_d_m_s;
+
+	// gps measurement matrix, measures position and velocity
+	math::Matrix<n_y_gps, n_x> C;
+	C(Y_gps_x, X_x) = 1;
+	C(Y_gps_y, X_y) = 1;
+	C(Y_gps_z, X_z) = 1;
+	C(Y_gps_vx, X_vx) = 1;
+	C(Y_gps_vy, X_vy) = 1;
+	C(Y_gps_vz, X_vz) = 1;
+
+	// gps covariance matrix
+	math::Matrix<n_y_gps, n_y_gps> R;
+	R(0,0) = _gps_xy_stddev.get()*_gps_xy_stddev.get();
+	R(1,1) = _gps_xy_stddev.get()*_gps_xy_stddev.get();
+	R(2,2) = _gps_z_stddev.get()*_gps_z_stddev.get();
+	R(3,3) = _gps_vxy_stddev.get()*_gps_vxy_stddev.get();
+	R(4,4) = _gps_vxy_stddev.get()*_gps_vxy_stddev.get();
+	R(5,5) = _gps_vz_stddev.get()*_gps_vz_stddev.get();
+
+	// residual
+	math::Matrix<6,6> S_I = ((C*_P*C.transposed()) + R).inversed();
+	math::Vector<6> r = y - C*_x;
+
+	// fault detection
+	float beta = sqrtf(r*(S_I*r));
+	if (beta > 3) { // 3 standard deviations away
+		if (!_gpsFault) {
+			mavlink_log_info(_mavlink_fd, "[lpe] gps fault, beta: %5.2f", double(beta));
+			warnx("[lpe] gps fault, beta: %5.2f", double(beta));
+		}
+		_gpsFault = 1;
+		// trust GPS less
+		S_I = ((C*_P*C.transposed()) + R*10).inversed();
+	} else if (_gpsFault) {
+		_gpsFault = 0;
+		mavlink_log_info(_mavlink_fd, "[lpe] GPS OK");
+		warnx("[lpe] GPS OK");
+	}
+
+	// kalman filter correction if no hard fault
+	if (_gpsFault < 2) {
+		math::Matrix<n_x, n_y_gps> K = _P*C.transposed()*S_I;
+		_x += K*r;
+		_P -= K*C*_P;
+	}
+	_time_last_gps = _timeStamp;
+}
+
+void BlockLocalPositionEstimator::correctVision() {
+
+	math::Vector<6> y;
+	y(0) = _sub_vision.get().x - _visionHome(0);
+	y(1) = _sub_vision.get().y - _visionHome(1);
+	y(2) = _sub_vision.get().z - _visionHome(2);
+	y(3) = _sub_vision.get().vx;
```

Comment:
About sync: we can do it on mavros with the timesync implementation you have done.